feat(splitChunks): support splitChunks.minSizeReduction (#9341)
web-infra-dev/rspack: website/docs/en/plugins/webpack/split-chunks-plugin.mdx

### splitChunks.minSize

#### splitChunks.cacheGroups.\{cacheGroup\}.minSize

- **Type:** `number | Record<string, number>`
- **Default:** `20000` in production and `10000` in others

When using the `number` type of configuration, the same `minSize` will be configured for all module types defined in [`splitChunks.defaultSizeTypes`](/plugins/webpack/split-chunks-plugin#splitchunksdefaultsizetypes).

```js title="rspack.config.mjs"
export default {
  //...
  optimization: {
    splitChunks: {
      minSize: 100 * 1000,
    },
  },
};
```

When configured with the object form, different `minSize` can be set for different types of module types defined in `splitChunks.defaultSizeTypes`.

```js title="rspack.config.mjs"
export default {
  //...
  optimization: {
    splitChunks: {
      minSize: {
        javascript: 100 * 1000,
        css: 300 * 1000,
      },
    },
  },
};
```

> For example, the above configuration means that the minimum size of javascript modules in the split chunks needs to be at least 100KB, and the minimum size of css modules needs to be at least 300KB.

Reviewer: ```suggestion
For example, the above configuration means that the minimum size of `javascript` modules in the split chunks needs to be at least 100KB, and the minimum size of `css` modules needs to be at least 300KB.
```

### splitChunks.minSize

#### splitChunks.cacheGroups.\{cacheGroup\}.minSize

- **Type:** `number | Record<string, number>`
- **Default:** `20000` in production and `10000` in others

Minimum size, in bytes, for a chunk to be generated.

### splitChunks.minSizeReduction

#### splitChunks.cacheGroups.\{cacheGroup\}.minSizeReduction

> - **Type: ** `number | Record<string, number>`

Reviewer: when should we use `Record<string, number>`, can you add an example too?

- **Default:** `0`

When output contains multiple small modules that collectively exceed the minSize threshold, developers may still want to prevent these modules from being split into separate chunks. The `minSizeReduction` parameter specifies the minimum total size reduction required for module splitting.

Assuming the following scenario, a module has a size of `40KB`, this module is owned by 2 chunks, with configuration `minSizeReduction: 100`, potential reduction in size for each parent chunk after splitting: `40KB`

Total reduction amount `40KB × 2 = 80KB`

> `80KB < 100KB` therefore no split is triggered

Reviewer: can we add a configuration example?

### splitChunks.minSize

#### splitChunks.cacheGroups.\{cacheGroup\}.minSize

- **Type:** `number | Record<string, number>`
- **Default:** `20000` in production and `10000` in others

Minimum size, in bytes, for a chunk to be generated.

### splitChunks.minSizeReduction

#### splitChunks.cacheGroups.\{cacheGroup\}.minSizeReduction

- **Type: ** `number | Record<string, number>`
- **Default:** `0`

> When output contains multiple small modules that collectively exceed the minSize threshold, developers may still want to prevent these modules from being split into separate chunks. The `minSizeReduction` parameter specifies the minimum total size reduction required for module splitting.
> 
> Assuming the following scenario, a module has a size of `40KB`, this module is owned by 2 chunks, with configuration `minSizeReduction: 100`, potential reduction in size for each parent chunk after splitting: `40KB`, total reduction amount `40KB × 2 = 80KB < 100KB`, therefore no split is triggered

Reviewer: ```suggestion
If there are several small modules in the build output, developers may not want to generate separate chunks for them even if their total size exceeds the `minSize` threshold. In this case, you can use the `minSizeReduction` parameter to set the minimum size reduction threshold required for module splitting.

The calculation rule for this parameter is: splitting will only occur when the total size reduction across all parent chunks after splitting the module is not less than the specified value.

Assuming the following scenario, suppose there is a 40KB module  that is referenced by 2 chunks, and we set `minSizeReduction: 100`. If we were to split this module, each parent chunk would be reduced by 40KB, resulting in a total reduction of `40KB × 2 = 80KB`. As this is less than 100KB, the split will not be triggered.
```

```js title="rspack.config.mjs"
export default {
  //...
  optimization: {
    splitChunks: {
      minSizeReduction: {
        javascript: 100 * 1000,

>         css: 100 * 1000,

Reviewer: What do javascript and css mean here? Can I use any string as the key?

Author: I rewrote the documentation for minSize and defaultSize, the previous one was very unclear, now I think it's easier to read?

Reviewer: LGTM 👍

### splitChunks.defaultSizeTypes

- **Type:** `string[]`
- **Default:** `["javascript", "unknown"]`, and if `experiments.css` is enabled, it will also include `"css"`

> When calculating the size of chunks, only the sizes of javascript modules and built-in css modules are taken into account by default. For example, when configuring `minSize: 300`, both javascript modules and css modules need to meet the requirement in order to be split.

Reviewer: ```suggestion
When calculating the size of chunks, only the sizes of `javascript` modules and built-in `css` modules are taken into account by default. For example, when configuring `minSize: 300`, both javascript modules and css modules need to meet the requirement in order to be split.
```
